appinventor/appengine/src/com/google/appinventor/server/LoginServlet.java

```diff
@@ -225,10 +247,80 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IO
       out.println("<p>" + bundle.getString("requestinstructions") + "</p>\n");
       out.println("<form method=POST action=\"" + req.getRequestURI() + "\">\n");
       out.println(bundle.getString("enteremailaddress") + ":&nbsp;<input type=text name=email value=\"\" size=\"35\"><br />\n");
+      out.println("<input type=hidden name=locale value=\"" + locale + "\">");
       out.println("<p></p>");
       out.println("<input type=submit value=\"" + bundle.getString("sendlink") + "\" style=\"font-size: 300%;\">\n");
       out.println("</form>\n");
       return;
+    } else if (page.equals("token")) {
+      String encodedToken = params.get("token");
+      if (encodedToken == null) {
+        fail(req, resp, "No Authentication Token Provided", locale);
+        return;
+      }
+      TokenProto.token token = null;
+      try {
+        token = Token.verifyToken(encodedToken);
+      } catch (TokenException e) {
+        fail(req, resp, e.getMessage(), locale);
+        return;
+      }
+      // At this point we have a valid token, so use it to login!
+      // need to make sure it is a SSOLOGIN token
+      if (token.getCommand() != TokenProto.token.CommandType.SSOLOGIN &&
+        token.getCommand() != TokenProto.token.CommandType.SSOLOGIN2 &&
+        token.getCommand() != TokenProto.token.CommandType.SSOLOGIN3) {
```

Comment:
Well, that isn't the case for this whole file. So I'm going to leave it be so it is consistent with the code around it. At some point we can fix the whole module, but I don't want to do that until we have the standalone/scaleable codebase all merged together. Otherwise I'll be facing the merge conflict from hades.

Reply:
Actually, I take that back, this is the only case, so I'm fixing it.